### PR TITLE
CI/33 Add dashboard skeleton and integration test

### PR DESCRIPTION
**Pull request by AlexAxthelm**
Comment action didn't trigger on this PR. See https://rmi-pacta.github.io/workflow.pacta.dashboard/pr-preview/pr-35/

Major changes:
- (CI) Adds Docker RMI checks + build process (`docker.yml`)
- (CI/Test) Adds "integration test" in running docker image (against outputs of `workflow.pacta`'s LATEST CI run, and publishing to GH pages (`gh-pages.yml`)
- Adds default dir: `dashboard_output_dir` for rendered dashboard files (dir can be served directly with `nginx`)
- Moves default dev data directory to `dashboard_output_dir/data`
- (CI) passes `hadolint`
- (R) adds function to build contents of `dashboard_output_dir`, and calling script updated to use that

Depends on https://github.com/RMI-PACTA/pacta-dashboard-svelte/pull/129

### CODE REVIEW

File: R/prepare_pacta_dashboard_data.R
Comment (jdhoffa) [inline R/prepare_pacta_dashboard_data.R:346]: Can you explain why you are disabling this? 
I know from a call we had that it was causing an error that it couldn't recover from, but can you please specify what the error was so that we have an idea of how to solve it

Comment (jdhoffa) [reply R/prepare_pacta_dashboard_data.R:346]: We do need this functionality, so I'd prefer we actually just address this now, rather than disable it

Comment (AlexAxthelm) [reply R/prepare_pacta_dashboard_data.R:346]: I don't have enough information to know quite what's wrong.

Full Docker run logs:
```
Digest: sha256:7b5a43eab6cedcd5fcb4c64d0b14d8e3e30b4b107bd73d9b2eb1560384254b7d
Status: Downloaded newer image for ghcr.io/rmi-pacta/workflow.pacta.dashboard:pr-35
Error in write_delim(x, file, delim = ",", na = na, append = append, col_names = col_names,  : 
  is.data.frame(x) is not TRUE
Calls: <Anonymous> ... zip_outputs -> %>% -> <Anonymous> -> write_delim -> stopifnot
Execution halted
```

https://github.com/RMI-PACTA/workflow.pacta.dashboard/actions/runs/12234692851/job/34124506400#step:7:93

Comment (AlexAxthelm) [reply R/prepare_pacta_dashboard_data.R:346]: I can address the functionality in a separate PR, once I've build up some telemetry around it (probably logging file names)

**Comment by jdhoffa** [comment]
@AlexAxthelm can you please name this PR more precisely, and it's intended function (for posterity, as this is what will appear in the git log and "integrate dashboard" isn't very descriptive)

File: .github/workflows/docker.yml
Comment (jdhoffa) [inline .github/workflows/docker.yml:36]: NIT: Consider removing commented out/ dead code before merging to `main`

File: .github/workflows/gh-pages.yml
Comment (jdhoffa) [inline .github/workflows/gh-pages.yml:11]: NIT: see above

File: .github/workflows/gh-pages.yml
Comment (jdhoffa) [inline .github/workflows/gh-pages.yml:70]: Unless these are left intentionally?

File: .github/workflows/gh-pages.yml
Comment (jdhoffa) [inline .github/workflows/gh-pages.yml:89]: In the spirit of keeping the `prod` and `dev` environments as closely aligned as possible, wouldn't it make more sense to just call the `docker-compose.yml` here? (otherwise, I'm not totally sure why we have a `docker-compose.yml` file)

https://12factor.net/dev-prod-parity

File: Dockerfile
Comment (jdhoffa) [inline Dockerfile:26]: I would name this something more descriptive... so far we have:
  dashboard_data_dir = Sys.getenv("DASHBOARD_DATA_DIR"),
  dashboard_files_dir = Sys.getenv("DASHBOARD_FILES_DIR"),
  dashboard_output_dir = Sys.getenv("DASHBOARD_OUTPUT_DIR")

Which is a bit confusing. Maybe `dashboard_build_files` would be more clear?

**Review by jdhoffa — CHANGES_REQUESTED** [review]
I have a few comments already!
Have a look and I'll do another review tomorrow 😊

File: .github/workflows/docker.yml
Comment (AlexAxthelm) [reply .github/workflows/docker.yml:36]: Hmm. most of this is going to come right back / be uncommented when I start work on #34, which is why I'm not too worried about it right now. But yeah, I can pull these out.

File: .github/workflows/gh-pages.yml
Comment (AlexAxthelm) [reply .github/workflows/gh-pages.yml:89]: The `docker-compose.yml` file is primarily a shortcut (to avoid having to type all the mounting and env flags). Additionally, that is (generally) set to `build: .`, rather than using a pre-built image (which is what we're testing here). 

We could add an override file with the CI settings, but passing the full image ID might start to get tricky there (unless we build an override file in the action).

the dev/prod parity issue isn't too much of a problem here, since we aren't using any backing services during execution, and all the backing services that we're using for environment setup (Azure Blobs) are inherently different than the ones we're using in PROD (Azure files, mounted to Container Instances), but we can't replicate that behavior on GHA.

**Review by jdhoffa — APPROVED** [review]

**Comment by jdhoffa** [comment]
> The tech mix plots on the portfolio-level and sector-level tabs of https://rmi-pacta.github.io/workflow.pacta.dashboard/pr-preview/pr-35/
> 
> are not working as expected
> 
> though they still work as expected here currently https://rmi-pacta.github.io/pacta-dashboard-svelte/
> 
> no idea what's causing that or if it matters, just pointing it out because it seems pretty weird

Indeed, totally weird!
However, it's important to consider that this PR no longer uses the typical `sample-data` found in the `pacta-dashboard-svelte` repo, but is pulling from Azure:

e.g.
``` yaml
          source: https://pactadatadev.blob.core.windows.net/ghactions-workflow-pacta-results/main/latest/full_params_2023Q4/analysis_output_dir
```
``` yaml
          source: https://pactadatadev.blob.core.windows.net/benchmarks-webapp/2023Q4/2023Q4_20240529T002355Z
```

etc.

**Comment by jdhoffa** [comment]
I will try to build and run the dashboard locally using these data and see if I can reproduce the behaviour

**Comment by jdhoffa** [comment]
This is totally related to: https://github.com/RMI-PACTA/workflow.pacta.dashboard/issues/27#issuecomment-2529577514

(and also the reason that the `zip_output` function had to be disabled), woohoo, CI is working!! 😊 
The short explanation is that, using the Azure input datasets, the output datasets `data_company_bubbles.json` and `data_techexposure.json` are empty.

This results in those plots not being able to be rendered.

**Comment by jdhoffa** [comment]
`data_techexposure.json` is empty because `start_year <- 2022`, however the earliest year in the input data is `2023`, so filters are removing all data. 

What's a bit bizarre however is that the `start_year` value is coming directly from the input data manifest:
https://github.com/RMI-PACTA/workflow.pacta.dashboard/blob/0ab208c3f09f3572d1d192a777327b50edc3df10/R/prepare_pacta_dashboard_data.R#L46

It's very surprising that those would be mis-aligned.

**Comment by jdhoffa** [comment]
> `data_techexposure.json` is empty because `start_year <- 2022`, however the earliest year in the input data is `2023`, so filters are removing all data.
> 
> What's a bit bizarre however is that the `start_year` value is coming directly from the input data manifest:
> 
> https://github.com/RMI-PACTA/workflow.pacta.dashboard/blob/0ab208c3f09f3572d1d192a777327b50edc3df10/R/prepare_pacta_dashboard_data.R#L46
> 
> It's very surprising that those would be mis-aligned.

Turns out it's the same issue for the `data_company_bubbles.json` dataset. Also filtering out all data when we filter `start_year == 2022`.

Now to understand why the `manifest.json` file is giving a `start_year` that is less than the earliest year in the analyzed data... 🤔

**Comment by jdhoffa** [comment]
This is so bizarre haha. Everything in the `manifest.json` file of the analysis outputs points to `2023` as expected, except for the `startYear`:
``` json
  "params": {
    "portfolio": {
      "holdingsDate": "2023-12-31",
      "files": ["default_portfolio.csv"],
      "name": "Default Portfolio"
    },
    "analysis": {
      "equityMarketList": ["GlobalMarket", "DevelopedMarket", "EmergingMarket"],
      "scenarioGeographiesList": ["Global", "GlobalAggregate", "NonOECD", "OECD"],
      "scenarioSourcesList": ["GECO2023", "ISF2023", "WEO2023"],
      "sectorList": ["Power", "Automotive", "Coal", "Steel", "Aviation", "Cement"],
      "startYear": 2022, -> this is funky for some reason
      "timeHorizon": 5
    }
  },
```

**Comment by jdhoffa** [comment]
Probably addressed by https://github.com/RMI-PACTA/workflow.pacta/issues/103

**Comment by AlexAxthelm** [comment]
Tech mix plots appear to be working, following merge of https://github.com/RMI-PACTA/workflow.pacta/pull/104 (fix start year in test params)

Screenshot_2024-12-10_09 28 28
Screenshot_2024-12-10_09 28 49

**Review by jdhoffa — APPROVED** [review]

**Pull request merged**
